peekingduck/viewer/viewer.py

```diff
@@ -0,0 +1,583 @@
+from peekingduck.viewer.viewer_gui import create_window
+from peekingduck.viewer.viewer_utils import (
+    get_keyboard_char,
+    get_keyboard_modifier,
+)
+
+####################
+# Globals
+####################
+BUTTON_DELAY: int = 250  # milliseconds (0.25 of a second)
+BUTTON_REPEAT: int = int(1000 / 60)  # milliseconds (60 fps)
+CHANGE_PIPELINE_DELAY: float = 1.0  # seconds before switching pipelines
+FPS_60: int = int(1000 / 60)  # milliseconds per iteration
+ZOOM_TEXT: List[str] = ["50%", "75%", "100%", "125%", "150%", "200%", "250%", "300%"]
+ZOOM_DEFAULT_IDX: int = 2
+ZOOMS: List[float] = [0.5, 0.75, 1.0, 1.25, 1.50, 2.00, 2.50, 3.00]  # > 3x is slow!
+EMOJI_PLAY = "\u25B6"
+EMOJI_STOP = "\u25FD"  # white medium-small square
+
+
+class Viewer:  # pylint: disable=too-many-instance-attributes, too-many-public-methods
+    """Implement PeekingDuck Viewer class"""
+
+    def __init__(
+        self,
+        pipeline_path: Path,
+        config_updates_cli: str,
+        custom_nodes_parent_subdir: str,
+        num_iter: int = 0,
+    ) -> None:
+        self.logger = logging.getLogger(__name__)
+        self.config_updates_cli = config_updates_cli
+        self.custom_nodes_parent_path = custom_nodes_parent_subdir
+        self.num_iter = num_iter
+        # init PlayList object
+        self.home_path = Path.home()
+        self.playlist = PlayList(self.home_path)
+        self.playlist.load_playlist_file()
+        self.playlist_show: bool = False
+        # init pipeline to run
+        self.init_pipeline(pipeline_path)
+        self.playlist.add_pipeline(self.pipeline_full_path)
+        # configure keyboard shortcuts map
+        self._keyboard_shortcuts = {
+            "z": self._zoom_reset,
+            "+": self._zoom_in,
+            "-": self._zoom_out,
+        }
+        # forward type hinting of internal working vars
+        self._frames: List[np.ndarray] = []
+        self._frame_idx: int = -1
+        self.zoom_idx: int = ZOOM_DEFAULT_IDX
+        self.is_output_playback: bool = False
+        self.is_pipeline_running: bool = False
+        self.state: str = "play"
+
+    def run(self) -> None:
+        """Main method to setup Viewer and run Tk event loop"""
+        self.logger.info(f"cwd={Path.cwd()}")
+        self.logger.info(f"pipeline={self.pipeline_path}")
+        # create Tkinter window and frames
+        create_window(self)
+        # trap macOS Cmd-Q keystroke
+        if platform.system() == "Darwin":
+            self.logger.info("binding macOS cmd-Q")
+            self.root.createcommand("::tk::mac::Quit", self.on_exit)
+        # activate internal timer function and start Tkinter event loop
+        self.timer_function()
+        self.root.mainloop()
+
+    ####################
+    #
+    # Tk Event Handlers
+    #
+    ####################
+    def btn_hide_show_playlist_press(self) -> None:
+        """Handle Hide/Show Playlist button
+
+        dotw technotes:
```

Comment:
Are the `dotw technotes` in various files personal notes that are meant to be removed eventually?

Reply:
The technotes are to document the important code/system behavior. Will remove `dotw` from the technotes, and change to `Tk technotes` where relevant.